### ISSUE TITLE
Duplicate events due to re-listing on eventrouter restart (update/crash)

### ISSUE THREAD

**Issue by hroumani** [heptiolabs/eventrouter#60]
Hello,

If the eventrouter controller were to be updated and/or crashes I believe it will re-drive the list operation on restart and correspondingly resend all the existing events (mind you by default k8 deletes events every hour I believe) ?

Assuming this is the case, while ideally the consumers are idempotent and can tolerate an at least once delivery semantic, it could potentially generate a lot of traffic if there are a lot of events.

Any thoughts on adding an annotation/label to the event "sent" on success to avoid this ?  Of course there's still the window of:
send (to whatever sink)
- crash
add annotation

But that's ok this is not meant for exactly once delivery but rather to handle more generic update/crash case and avoid a lot of resends.

I could see some hesitation on mucking with the k8 event though.

Any thoughts ?

**Comment by timothysc** [comment]
Yes, I wanted to add a resource version field to a config map.  The problem with that is finding the right write frequency.  If I write on every event that would also be too spammy.

**Comment by hroumani** [comment]
Thanks @timothysc !  Trying to grok that though - by version do you mean the config map would contain a tuple <eventid,last_published_version> for every event and you would check that before publishing to your sink ?

Assuming that's an accurate description (I apologise if I misinterpreted it) that would potentially be a lot of data to store in the config map.  Would the annotation approach/label approach be more scalable (i.e. writing to the resource/obj it self in etcd) is the concern making changes to that object ?

**Comment by timothysc** [comment]
resource version of the event, and apply that to the next watch.

**Comment by jonmoter** [comment]
Regarding frequency, could you make that a config flag or ENV var with a sensible default?  Like write a checkpoint every N seconds or N events, and right at termination?

Seems like that's a trade-off between the number of extra ConfigMap updates you want to do, vs how much duplicate data you're willing to tolerate if the process crashes without cleaning up.
